src/waveform/renderers/allshader/digitsrenderer.cpp

```diff
@@ -0,0 +1,196 @@
+#include "waveform/renderers/allshader/digitsrenderer.h"
+
+#include <QColor>
+#include <QFontMetricsF>
+#include <QGraphicsBlurEffect>
+#include <QGraphicsPixmapItem>
+#include <QGraphicsScene>
+#include <QOpenGLTexture>
+#include <QPainter>
+#include <QPainterPath>
+
+#include "waveform/renderers/allshader/matrixforwidgetgeometry.h"
+#include "waveform/renderers/allshader/vertexdata.h"
+
+// Render digits using a texture (generated) with digits with blurred dark outline
+
+namespace {
+
+// The texture will contain 12 characters: 10 digits, colon and dot
+constexpr int NUM_CHARS = 12;
+
+// space around chars for blurred dark outline
+constexpr qreal SPACE = 3;
+// extra kerning when rendering chars
+constexpr qreal KERNING = 1;
+
+char indexToChar(int index) {
+    const char* str = "0123456789:.";
+    return str[index];
+}
+int charToIndex(char ch) {
+    if (ch >= '0' && ch <= '9') {
+        return static_cast<int>(ch - '0');
+    }
+    if (ch == ':') {
+        return 10;
+    }
+    if (ch == '.') {
+        return 11;
+    }
+    assert(false);
+    return 11; // fallback to dot
+}
+} // namespace
+
+allshader::DigitsRenderer::~DigitsRenderer() = default;
+
+void allshader::DigitsRenderer::init() {
+    initializeOpenGLFunctions();
+    m_shader.init();
+}
+
+float allshader::DigitsRenderer::height() const {
+    return static_cast<float>(m_texture.height());
+}
+
+void allshader::DigitsRenderer::generateTexture(int fontPixelSize, float devicePixelRatio) {
+    QFont font;
+    font.setFamily("Open Sans");
+    font.setPixelSize(fontPixelSize);
+
+    QFontMetricsF metrics{font};
+
+    qreal totalTextWidth = 0;
+    qreal maxTextHeight = 0;
+    for (int i = 0; i < NUM_CHARS; i++) {
+        assert(charToIndex(indexToChar(i)) == i);
```

Comment:
wdyt about making both of these constexpr and then assert it for the entire valid set of characters in a static_assert instead?

Reply:
not sure I understand what you mean

Reply:
something like this: 
```
constexpr bool checkCharToIndex() {
    for (int i = 0; i < NUM_CHARS; i++) {
        if (charToIndex(indexToChar(i)) != i) {
        	return false;
        }
    }
    return true;
}
static_assert(checkCharToIndex());
```

charToIndex() and indexToChar() needs to become constexpr as well see my inline comments.